cpu/cc2538/lpm.c

```diff
@@ -212,13 +222,40 @@ lpm_exit()
    * We need to convert sleep duration from rtimer ticks to clock ticks */
   clock_adjust();
 
-  /* Restore system clock to the 32 MHz XOSC */
+  /* CC2538 silicon errata:
+     Possible Incorrect Value of Clock Dividers after PM2 and PM3 */
+#if SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ || \
+  SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
+  if(reset_div_mask) {
+    /* We come from PM2 or PM3 where we adjusted (=set to 16 MHz)
+       either one or both divisors according to the errata note.
+       Set them back to 32 MHz. */
+    uint32_t val = REG(SYS_CTRL_CLOCK_CTRL);
+
+#if SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ && \
+    SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
+    /* Set both divisors back to 0 (=32MHz) */
+    val &= ~(SYS_CTRL_CLOCK_CTRL_SYS_DIV + SYS_CTRL_CLOCK_CTRL_IO_DIV);
+#elif SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ
+    /* Set sys clock divisor back to 0 (=32MHz) */
+    val &= ~SYS_CTRL_CLOCK_CTRL_SYS_DIV;
+#elif SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
+    /* Set io clock divisor back to 0 (=32MHz) */
+    val &= ~SYS_CTRL_CLOCK_CTRL_IO_DIV;
+#endif
+
+    REG(SYS_CTRL_CLOCK_CTRL) = val;
+    reset_div = 0;
+  }
+#endif
+
+/* Restore system clock to the 32 MHz XOSC */
   select_32_mhz_xosc();
```

Comment:
After second thought, it's better to leave this unchanged and not to swap `clock_adjust()` and `select_32_mhz_xosc()` because this could not actually improve anything, but this could increase the wake-up latency due to the reduced time before triggering the switch to the crystal oscillator, so the switching time would be spent doing nothing but waiting instead of being used by `clock_adjust()`.